### ISSUE TITLE
Mounting whole IPFS at once

### ISSUE THREAD

**Issue by SupraSummus** [SupraSummus/ipfs-api-mount#33]
Currently ipfs-api-mount mounts one specified CID under one specified directory.

For some use cases it may be convenient to mount whole IPFS namespace in that given directory, similarly, as it's done by `ipfs mount` in go-ipfs implementation.

**Comment by SupraSummus** [comment]
I'm not sure posix filesystem can express semantics of whole IPFS. Let's suppose we mount whole ipfs under `/ipfs`. Then we do `ls /ipfs/QmSomeHash`. `ipfs-api-mount` does a request to ipfs daemon to list the object, but timeout passes. What should we do then?
1. Return ENOENT (No such file or directory). This is correct assuming timeout passed because there is no directory with that hash. But we might just has a network hiccup and the object exists. We cannot get it in time.
2. Return EAGAIN (Resource temporarily unavailable). This is correct assuming we didn't have enough time to retrieve the object. But the object may be unavailable in a little longer time span than "temporary". Maybe it's a porn that will be recorded on Mars and uploaded 2302 A.D.? Or maybe that object is completely nonexistent, because the hash cannot be achieved by any IPLD object.

None of the solutions I proposed is good. If I had to go, I'd choose 2 though.

There is also EWOULDBLOCK - defined as "Operation would block". It's a good candidate, but it's often the same value as EAGAIN (or maybe it's irrelevant that it's the same).

**Comment by SupraSummus** [comment]
@jcea described his use case:
> I also usually use IPFS to interface with several unrelated different contents. Having everything under "/ipfs/" is very convenient. Needing to kill "ipfs-api-mount" instance and launch a new one for each source I am interacting with is so inconvenient that I am still using "ipfs mount" most of the time. It is slow, but far more practical and less demanding for my mind.
> 
> Your current approach is practical when you are interacting with a single static source. That is not my use case. I interact with multiple sources and most of the time they are dynamic ("ipfs files", for instance). That situation makes this program unsuitable for me, most of the time.

So there are two issues:
1. Interacting with many root hashes at once.

   Mounting whole IPFS surely would provide convenience, but I think it's a harmful shortcut (because unix FS operations can't express every possible scenario we can encounter mounting whole IPFS). I feel like interacting with multiple trees can be achieved in an elegant manner using multiple ipfs-api-mount instances. Or by wrapping desired root hashes in single object and then mounting wrapper object. My guess is that it can be automated and used with ease. If I would do it I would start by trying autofs, its wildcard mount feature, and timed unmount.

2. Interacting with changing resources. As I understand it @jcea proposes to `cd` into right `mountpoint/QmSomeHash` every time we want the content of the filesystem to be updated.

   Mounting whole ipfs won't provide dynamic resources. There will be many of them, and `cd mountpoint/QmSomeHash` will be easy, but this is not dynamic. Any file open under `mountpoint/QmSomeHash` wont be affected by working dir change. To truly support changing resources we need remount feature.

**Comment by SupraSummus** [comment]
One more against: remount feature is incompatibile with mounting whole ipfs. If we mount whole IPFS we cant substitute any tree, because all of them contain hash in the path, which ensures immutability. Afaik we can't go with links that will switch target from `mountpoint/QmOldHash` to `mountpoint/QmNewHash`, because that won't update already opened files.

**Comment by SupraSummus** [comment]
It looks I'm very much biased against "total IPFS mount" ;-D @jcea help me break this if you would

**Comment by jcea** [comment]
> I'm not sure posix filesystem can express semantics of whole IPFS. Let's suppose we mount whole ipfs under `/ipfs`. Then we do `ls /ipfs/QmSomeHash`. `ipfs-api-mount` does a request to ipfs daemon to list the object, but timeout passes. What should we do then?
> 
>     1. Return ENOENT (No such file or directory). This is correct assuming timeout passed because there is no directory with that hash. But we might just has a network hiccup and the object exists. We cannot get it in time.
> 
>     2. Return EAGAIN (Resource temporarily unavailable). This is correct assuming we didn't have enough time to retrieve the object. But the object may be unavailable in a little longer time span than "temporary". Maybe it's a porn that will be recorded on Mars and uploaded 2302 A.D.? Or maybe that object is completely nonexistent, because the hash cannot be achieved by any IPLD object.
> 
> 
> None of the solutions I proposed is good. If I had to go, I'd choose 2 though.
> 
> There is also EWOULDBLOCK - defined as "Operation would block". It's a good candidate, but it's often the same value as EAGAIN (or maybe it's irrelevant that it's the same).

Note that this issue is happening already. You can mount a CID and it or its children could be unavailable.

NFS semantic says "wait forever", but since IPFS is flacky and there is not guarantee of the resource being available  ever, I would actually return an EIO error after a configurable timeout.

EAGAIN is risky because some programming languages could just try again... and again... and again without notifying the error to upper layers.

In any case, this is not related to this enhancement. The "want to do when we can not get a IPFS block" is already present in current implementation. For instance:

```
jcea@jcea:/tmp$ ipfs-api-mount /ipfs/bafybeib6x5a7uytkyq6nnfl42bmhcnjroxfzecfmjiurmztzyfliejr6um /tmp/a
```

It sits there... forever.

If a mount success but any children is unavailable, "ipfs-api-mount" just will keep there trying forever.

**Comment by jcea** [comment]
I am not talking about remounting anything. The point of mounting the complete ipfs hierarchy is to avoid remounting.

Please, explain how could you interact with a "ipfs files" mutable environment (one of my usecases) with the "use several instances of ipfs-api-mount". Any change will create a new "root". Do you really think I would launch a new mount instance per file change? You just can't.

I think you have a misconception somewhere. Mounting the entire hierarchy doesn't requires any kind of remount and previous "open files" are not affected.

I think you are misundertanding that IPFS mount would do. Since the entire IPFS universe is sitting there, your open files from anywhere. When I move around using "cd", I am changing the directory I am caring about now, but all the entire hierarchy is available and any open file is just there.

Let's think about regular Unix filesystem: In Python, I could do ```open("/etc/passwd")``` even when my current working directory is ```/home/jcea/```. I can do too ```f=open("my_file.txt")```, then change current working directory and open ```f``` file doesn't care AT ALL.

On other words, being ANYWHERE in "/ipfs/AAA" (current working directory), I could be able to do ```f=open("/ipfs/BBBB/random.file")```. After opening that file, I could change my current directory to ```/ipfs/CCC``` and ```f``` filehandle doesn't care at all.

The only "strange" thing is that "/ipfs" modes would be "--x" (traverse only)" because a) it is a read only filesystem and b) you can not just list all content. But if you know the name (the CID) of the node you are interested in, you can just open it with ```open("/ipfs/CID")``` whatever is your current working directory.
Think this way: mount the entire IPFS hierarchy is the same that a regular unix filesystem. You can open files anywhere, in different directories while your current directory is whatever. Your current approach would be more similar to "chroot" syscall: Take the complete ipfs hierarchy and limit yourself to a small subtree.

**Comment by jcea** [comment]
> One more against: remount feature is incompatibile with mounting whole ipfs. If we mount whole IPFS we cant substitute any tree, because all of them contain hash in the path, which ensures immutability. Afaik we can't go with links that will switch target from `mountpoint/QmOldHash` to `mountpoint/QmNewHash`, because that won't update already opened files.

I don't understand why you are insisting about "remounting". When I open a file in IPFS, I am opening a particular revision of it. Since IPFS is inmutable, a file change will create a new file or, let's say, a new revision of the old file. Of course, if I have a open file to revision X of a file, I expect that filehandle to keep referring to that particular revision. It would be highly surprising that the open file changed silently under my nose.

This is not different to regular unix: If you open a file, leave it open, delete the file and create a new file with the same name, the old filehandle keeps referring to the old (deleted) file. I can seek/read/write on it. The old file will be available in a ghost state until the old filehandle is closed. It doesn't matter that there is a new file with the same name around there.

**Comment by SupraSummus** [comment]
@jcea I've implemented mounting whole ipfs in #38 as a separate script (`ipfs-api-mount-whole`). Feel free to try it out and share your thoughts.

I'm still not convinced that mounting whole IPFS is a good idea in the long run, but I see that people got used to `ipfs mount` functionality and not providing it in ipfs-api-mount may be seen as a lack of crucial functionality.

@jcea I'm planning to reply to your arguments for mounting whole IPFS, but right now I've got no time/will to translate my thinking into english ;) So for now I'll leave them hanging.

**Comment by SupraSummus** [comment]
> Note that this issue is happening already. You can mount a CID and it or its children could be unavailable.

Yes, the issue is happening, but when it's happening it's because of actual malfunction of underlying layers or because of filesystem misconfiguration (passing corrupted root path). Both of these causes are unavoidable in every non-trivial filesystem. We should do as much as possible to handle them properly, but if we can't that's okay, because in general the situation may be unrecoverable.

When mounting whole IPFS, and user is requesting mountpoint/QmSomeHash that can't be retrieved we get different scenario. Filesystem is configured alright and there is no malfunction. In that scenario we MUST handle every filesystem request properly. Handling properly requires returning proper errno when error happens. EAGAIN, ENOENT are not suitable. EPERM is not suitable (because whole IPFS is immutable, so permissions can't change). EIO is unsuitable because there was no I/O error. ETIMEDOUT may be okay however.

> Please, explain how could you interact with a "ipfs files" mutable environment (one of my usecases) with the "use several instances of ipfs-api-mount". Any change will create a new "root". Do you really think I would launch a new mount instance per file change? You just can't.

With mfs I won't go with creating new mount. I might not been clear enough - multiple mounts are good when you have multiple trees that you want to access at once. When you want to "checkout" to different tree my intuition is to implement and use "remount".

So I imagine with whole ipfs mounted you would do something like this:

```
ipfs-api-mount-whole mountpoint
# do some operations with `ipfs files`
cd mountpoint/$(ipfs files stat --hash /) 
```

With remount feature (that is not implemented yet) you could do something like

```
mkfifo hashes
cat hashes | ipfs-api-mount --read-hashes-from-stdin QmInitialHash mountpoint
# interact with `ipfs files`
ipfs files stat --hash / > hashes
```

> Let's think about regular Unix filesystem: In Python, I could do open("/etc/passwd") even when my current working directory is /home/jcea/. I can do too f=open("my_file.txt"), then change current working directory and open f file doesn't care AT ALL.

Exactly. Imagine a program that doesn't know anything about IPFS. It expects just a unix filesystem, and even a mutable one. For example let's say it's a `tail --follow=descriptor mountpoint/name`. It will open file just once and then print any new data that is appended to it. With remount functionality you can simulate changing filesystem, such a program will understand. That's a feature you can't do when mounting whole IPFS. Or am I wrong?

> This is not different to regular unix: If you open a file, leave it open, delete the file and create a new file with the same name, the old filehandle keeps referring to the old (deleted) file. I can seek/read/write on it. The old file will be available in a ghost state until the old filehandle is closed. It doesn't matter that there is a new file with the same name around there.

Deleting files works just like you described. Modifying files is quite different story. In a "regular unix fs" writing to a file won't create a new file. Other programs that have the file open can see the change right away, without reopening. To provide something like this in ipfs-api-mount we need remounting (or am I wrong again?).
